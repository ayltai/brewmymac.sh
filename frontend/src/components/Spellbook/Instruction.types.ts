export interface InstructionProps {
    open?            : boolean,
    spellbookId      : string,
    onClose?         : () => void,
    [ key : string ] : any,
}
