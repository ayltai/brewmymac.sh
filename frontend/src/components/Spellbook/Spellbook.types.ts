export interface SpellbookProps {
    [ key : string ] : any,
}
