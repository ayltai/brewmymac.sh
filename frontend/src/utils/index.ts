import * as Sentry from '@sentry/react';
import { FirebaseApp, } from 'firebase/app';
import { getAnalytics, logEvent, } from 'firebase/analytics';

export const handleError = (error : any) => {
    if (import.meta.env.PROD) {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const logPageView = (app : FirebaseApp, {
    pageName,
    route,
} : {
    pageName : string,
    route    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'page_view', {
            page_title    : pageName,
            page_location : route,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logAddToCart = (app : FirebaseApp, {
    itemId,
} : {
    itemId : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'add_to_cart', {
            items : [
                {
                    item_id : itemId,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logRemoveFromCart = (app : FirebaseApp, {
    itemId,
} : {
    itemId : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'remove_from_cart', {
            items : [
                {
                    item_id : itemId,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logCheckout = (app : FirebaseApp, {
    itemIds,
} : {
    itemIds : string[],
}) => {
    try {
        logEvent(getAnalytics(app), 'checkout', {
            items : itemIds.map((itemId : string) => ({
                item_id : itemId,
            })),
        });
    } catch (error) {
        handleError(error);
    }
};

export const logSearch = (app : FirebaseApp, {
    query,
} : {
    query : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'search', {
            search_term : query,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logShare = (app : FirebaseApp, {
    product,
    itemId,
} : {
    product : 'packages' | 'tweaks',
    itemId  : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'share', {
            content_type : product,
            item_id      : itemId,
        });
    } catch (error) {
        handleError(error);
    }
};
