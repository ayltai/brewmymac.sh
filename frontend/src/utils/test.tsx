import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, Store, } from '@reduxjs/toolkit';
import { render, } from '@testing-library/react';
import React, { ReactNode, } from 'react';
import { Provider, } from 'react-redux';
import { MemoryRouter, } from 'react-router-dom';

import { appStoreApi, homebrewApi, } from '../apis';
import { cookbookReducer, preferenceReducer, spellbookReducer, } from '../states';
import { appTheme, } from '../styles';

export const createStore = (preloadedState? : unknown) => configureStore({
    preloadedState,
    reducer    : {
        cookbook                    : cookbookReducer,
        preference                  : preferenceReducer,
        spellbook                   : spellbookReducer,
        [ appStoreApi.reducerPath ] : appStoreApi.reducer,
        [ homebrewApi.reducerPath ] : homebrewApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(appStoreApi.middleware, homebrewApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

// @ts-ignore
const theme = createTheme(appTheme);

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </MemoryRouter>
        </Provider>
    ),
    ...rest,
});

// eslint-disable-next-line react-refresh/only-export-components
export * from '@testing-library/react';

export { customRender as render, };
