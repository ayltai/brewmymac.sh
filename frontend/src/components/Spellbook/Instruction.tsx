import Box from '@mui/material/Box';
import Link from '@mui/material/Link';
import Stack from '@mui/material/Stack';
import Typography from '@mui/material/Typography';
import React, { FC, useContext, } from 'react';
import { useTranslation, } from 'react-i18next';

import { API_ENDPOINT, } from '../../constants';
import { FirebaseContext, } from '../../contexts';
import { useAppSelector, } from '../../hooks';
import type { Tweak, } from '../../models';
import { logShare, } from '../../utils';
import { CopyableInput, } from '../CopyableInput';
import { DialogView, } from '../DialogView';
import type { InstructionProps, } from './Instruction.types';

/**
 * An instruction dialog that informs users about how to apply the selected tweaks.
 * @param open Whether the dialog is open
 * @param spellbookId The ID of the spellbook
 * @param onClose Function to call when the close button is clicked
 * @param rest Other props
 */
export const Instruction : FC<InstructionProps> = ({
    open,
    spellbookId,
    onClose,
    ...rest
}) => {
    const { tweaks,    } : { tweaks    : Tweak[],          } = useAppSelector(state => state.spellbook);
    const { themeMode, } : { themeMode : 'light' | 'dark', } = useAppSelector(state => state.preference);

    const app  = useContext(FirebaseContext);

    const { t, } = useTranslation();

    const handleCopy = () => {
        if (app) logShare(app, {
            product : 'packages',
            itemId  : spellbookId,
        });
    };

    return (
        <DialogView
            open={open}
            title={t('tweaks.checkout.title')}
            onClose={onClose}
            {...rest}>
            <Stack spacing={1}>
                {tweaks.some(tweak => tweak.sipInvolved) && (
                    <>
                        <Box paddingBottom={2}>
                            <Typography>
                                {t('tweaks.checkout.instructions.1')}
                            </Typography>
                            <Box
                                sx={{
                                    borderRadius    : '1rem',
                                    backgroundColor : themeMode === 'light' ? '#fbe9e7' : '#bf360c',
                                }}
                                padding={2}>
                                {tweaks.filter(tweak => tweak.sipInvolved).map(tweak => (
                                    <Typography key={tweak.id}>
                                        {`- ${tweak.name}`}
                                    </Typography>
                                ))}
                            </Box>
                        </Box>
                        <Box paddingBottom={2}>
                            <Typography>
                                {t('tweaks.checkout.instructions.2')}
                            </Typography>
                            <Link
                                href='https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection'
                                target='_blank'>
                                https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection
                            </Link>
                        </Box>
                    </>
                )}
                <Typography>
                    {t('tweaks.checkout.copy')}
                </Typography>
                <Link
                    href='https://support.apple.com/en-gb/guide/terminal/apd5265185d-f365-44cb-8b09-71a064a42125/mac'
                    target='_blank'>
                    https://support.apple.com/en-gb/guide/terminal/apd5265185d-f365-44cb-8b09-71a064a42125/mac
                </Link>
                <CopyableInput
                    value={`bash -c "$(curl -fsSL '${import.meta.env.DEV ? `http://${window.location.hostname}:8080` : API_ENDPOINT}/spellbooks/install/${spellbookId}')"`}
                    onCopy={handleCopy} />
            </Stack>
        </DialogView>
    );
};
