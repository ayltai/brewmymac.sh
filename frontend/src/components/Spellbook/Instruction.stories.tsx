import type { Meta, StoryObj, } from '@storybook/react';
import React from 'react';
import { Provider, } from 'react-redux';

import { createStore, } from '../../utils/test';
import { SpellbookInstruction, } from './Instruction';

const meta : Meta<typeof SpellbookInstruction> = {
    component : SpellbookInstruction,
    tags      : [
        'autodocs',
    ],
    argTypes  : {
        transactionId : {
            description : 'The ID of the transaction',
        },
    },
};

export default meta;

type Story = StoryObj<typeof SpellbookInstruction>;

const store = createStore({
    spellbook : {
        tweaks : [
            {
                id          : '1',
                name        : 'magic.1',
                description : 'Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor.',
                author      : 'BrewMyMac',
                parameters  : [
                    'Disable',
                ],
                types       : [
                    'boolean',
                ],
                values      : [
                    true,
                ],
            },
            {
                id          : '2',
                name        : 'magic.2',
                description : 'Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor.',
                author      : 'BrewMyMac',
                parameters  : [
                    'Name',
                    'Width',
                    'Height',
                ],
                types       : [
                    'string',
                    'number',
                    'number',
                ],
                values      : [
                    'Magic',
                    1920,
                    1024,
                ],
                sipInvolved : true,
            },
        ],
    },
});

export const Default : Story = {
    args : {
        open          : true,
        transactionId : '1',
    },
    decorators : [
        StoryComponent => (
            <Provider store={store}>
                <StoryComponent />
            </Provider>
        ),
    ],
};
