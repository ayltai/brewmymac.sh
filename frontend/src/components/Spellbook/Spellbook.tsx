import React, { useCallback, useEffect, useState, } from 'react';

import { useAddSpellbookMutation, } from '../../apis';
import { useAppDispatch, useAppSelector, } from '../../hooks';
import type { Tweak, } from '../../models';
import { removeTweak, updateTweak, } from '../../states';
import { handleError, } from '../../utils';
import { TweakDetailsView, } from '../DetailsView';
import { ShoppingCartDetail, } from '../ShoppingCartDetail';
import { SpellbookInstruction, } from './Instruction';

/**
 * A spellbook contains a list of selected tweaks.
 */
export const Spellbook = () => {
    const [ openInstruction, setOpenInstruction, ] = useState(false);
    const [ selected,        setSelected,        ] = useState<Tweak | undefined>();
    const [ spellbookId,     setSpellbookId,     ] = useState<string>();

    const { tweaks, } : { tweaks : Tweak[], } = useAppSelector(state => state.spellbook);

    const [ addSpellbook, { data, error, reset, }, ] = useAddSpellbookMutation();

    const dispatch = useAppDispatch();

    const handleSelect = (item : Tweak) => setSelected(item);

    const handleDelete = (item : Tweak) => dispatch(removeTweak(item));

    const handleNext = () => setOpenInstruction(true);

    const handleChange = useCallback((parameter : string, value? : boolean | number | string) => {
        if (selected && value) {
            const tweak = {
                ...selected,
            };

            tweak.values[selected.parameters.indexOf(parameter)] = value;

            dispatch(updateTweak(tweak));
        }
    }, [ dispatch, selected, ]);

    const handleCloseInstruction = () => setOpenInstruction(false);

    useEffect(() => {
        if (openInstruction && tweaks.length > 0) addSpellbook(tweaks).catch(handleError);
    }, [ addSpellbook, tweaks, openInstruction, ]);

    useEffect(() => {
        if (data) setSpellbookId(data);
    }, [ data, ]);

    useEffect(() => {
        if (error) handleError(error);
    }, [ error, ]);

    return (
        <ShoppingCartDetail
            items={tweaks}
            detailsView={
                <TweakDetailsView
                    description={selected?.description}
                    infoUrl={selected?.infoUrl}
                    parameters={selected?.parameters ?? []}
                    types={selected?.types ?? []}
                    values={selected?.values ?? []}
                    onChange={handleChange} />
            }
            instruction={
                <SpellbookInstruction
                    open={openInstruction}
                    transactionId={spellbookId}
                    onClose={handleCloseInstruction} />
            }
            instructionShown={openInstruction}
            onSelect={handleSelect}
            onDelete={handleDelete}
            onNext={handleNext}
            checkout={addSpellbook}
            reset={reset} />
    );
};
