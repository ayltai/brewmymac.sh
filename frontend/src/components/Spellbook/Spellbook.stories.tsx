import type { Meta, StoryObj, } from '@storybook/react';
import React from 'react';
import { Provider, } from 'react-redux';

import { createStore, } from '../../utils/test';
import { Spellbook, } from './Spellbook';

const meta : Meta<typeof Spellbook> = {
    component : Spellbook,
    tags      : [
        'autodocs',
    ],
    argTypes  : {
    },
};

export default meta;

type Story = StoryObj<typeof Spellbook>;

const store = createStore({
    spellbook : {
        tweaks : [
            {
                id          : '1',
                name        : 'magic.1',
                description : 'Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor.',
                author      : 'BrewMyMac',
                parameters  : [
                    'Disable',
                ],
                types       : [
                    'boolean',
                ],
                values      : [
                    true,
                ],
            },
            {
                id          : '2',
                name        : 'magic.2',
                description : 'Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor.',
                author      : 'BrewMyMac',
                parameters  : [
                    'Name',
                    'Width',
                    'Height',
                ],
                types       : [
                    'string',
                    'number',
                    'number',
                ],
                values      : [
                    'Magic',
                    1920,
                    1024,
                ],
            },
        ],
    },
});

export const Default : Story = {
    args       : {
    },
    decorators : [
        StoryComponent => (
            <Provider store={store}>
                <StoryComponent />
            </Provider>
        ),
    ],
};
