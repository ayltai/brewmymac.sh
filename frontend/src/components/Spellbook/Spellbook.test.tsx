import React from 'react';

import { render, } from '../../utils/test';
import { Spellbook, } from './Spellbook';

describe('<Spellbook />', () => {
    it('renders correctly', () => expect(render(
        <Spellbook />,
        {
            spellbook : {
                tweaks : [
                    {
                        id          : '1',
                        name        : 'magic.1',
                        description : 'Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor.',
                        author      : 'BrewMyMac',
                        parameters  : [
                            'Disable',
                        ],
                        types       : [
                            'boolean',
                        ],
                        values      : [
                            'true',
                        ],
                    },
                    {
                        id          : '2',
                        name        : 'magic.2',
                        description : 'Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor.',
                        author      : 'BrewMyMac',
                        parameters  : [
                            'Name',
                            'Width',
                            'Height',
                        ],
                        types       : [
                            'string',
                            'number',
                            'number',
                        ],
                        values      : [
                            'Magic',
                            1920,
                            1024,
                        ],
                    },
                ],
            },
        }
    )).toMatchSnapshot());
});
