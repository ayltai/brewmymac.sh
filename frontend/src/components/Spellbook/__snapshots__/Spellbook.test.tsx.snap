// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Spellbook /> > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-1m6lmqt-MuiStack-root"
      >
        <div
          class="MuiStack-root css-1sf5043-MuiStack-root"
        >
          <div
            class="MuiBox-root css-i9gxme"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-f70x3d-MuiSvgIcon-root"
            data-testid="AutoFixOffIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m23 1-2.5 1.4L18 1l1.4 2.5L18 6l2.5-1.4L23 6l-1.4-2.5zm-8.34 6.22 2.12 2.12-2.44 2.44.81.81 2.55-2.55c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0L11.4 8.84l.81.81zm-.78 6.65-3.75-3.75-6.86-6.86L2 4.53l6.86 6.86-6.57 6.57c-.39.39-.39 1.02 0 1.41l2.34 2.34c.39.39 1.02.39 1.41 0l6.57-6.57L19.47 22l1.27-1.27z"
            />
          </svg>
          <h5
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1b7t60j-MuiTypography-root"
          >
            cart.empty
          </h5>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
          >
            cart.hint
          </p>
          <div
            class="MuiBox-root css-i9gxme"
          />
        </div>
        <div
          aria-hidden="true"
          class="MuiBackdrop-root css-sox5kk-MuiBackdrop-root"
          style="opacity: 0; visibility: hidden;"
        >
          <div
            class="MuiStack-root css-6rg9iv-MuiStack-root"
          >
            <span
              class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-mod6k9-MuiCircularProgress-root"
              role="progressbar"
              style="width: 40px; height: 40px;"
            >
              <svg
                class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                viewBox="22 22 44 44"
              >
                <circle
                  class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                  cx="44"
                  cy="44"
                  fill="none"
                  r="20.2"
                  stroke-width="3.6"
                />
              </svg>
            </span>
            cart.processing
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-1m6lmqt-MuiStack-root"
    >
      <div
        class="MuiStack-root css-1sf5043-MuiStack-root"
      >
        <div
          class="MuiBox-root css-i9gxme"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-f70x3d-MuiSvgIcon-root"
          data-testid="AutoFixOffIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m23 1-2.5 1.4L18 1l1.4 2.5L18 6l2.5-1.4L23 6l-1.4-2.5zm-8.34 6.22 2.12 2.12-2.44 2.44.81.81 2.55-2.55c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0L11.4 8.84l.81.81zm-.78 6.65-3.75-3.75-6.86-6.86L2 4.53l6.86 6.86-6.57 6.57c-.39.39-.39 1.02 0 1.41l2.34 2.34c.39.39 1.02.39 1.41 0l6.57-6.57L19.47 22l1.27-1.27z"
          />
        </svg>
        <h5
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1b7t60j-MuiTypography-root"
        >
          cart.empty
        </h5>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
        >
          cart.hint
        </p>
        <div
          class="MuiBox-root css-i9gxme"
        />
      </div>
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-sox5kk-MuiBackdrop-root"
        style="opacity: 0; visibility: hidden;"
      >
        <div
          class="MuiStack-root css-6rg9iv-MuiStack-root"
        >
          <span
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-mod6k9-MuiCircularProgress-root"
            role="progressbar"
            style="width: 40px; height: 40px;"
          >
            <svg
              class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </span>
          cart.processing
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
